core/domain/classifier_services_test.py

```diff
@@ -623,3 +630,162 @@ def test_can_not_store_classifier_data_due_to_invalid_job_id(self):
+            classifier_services.get_state_training_jobs_mapping(
+                exp_id, 1, invalid_state_name))
+        self.assertIsNone(state_training_jobs_mapping)
+
+    def test_migrate_state_training_jobs(self):
+        """Test the migrate_state_training_jobs method."""
+
+        state_name = 'Home'
+        mock_interaction_classifier_mapping = {
+            'TextInput': {
+                'algorithm_id': 'NewTextClassifier',
+                'algorithm_version': 1
+                },
+            }
+        with self.swap(
+            feconf,
+            'INTERACTION_CLASSIFIER_MAPPING',
+            mock_interaction_classifier_mapping):
+            classifier_services.migrate_state_training_jobs(
+                classifier_services.get_state_training_jobs_mapping(
+                    self.exp_id, 1, state_name)
+             )
+        state_training_jobs_mapping = (
+            classifier_services.get_state_training_jobs_mapping(
+                self.exp_id, 1, 'Home')
+            )
+        self.assertIn(
+            'NewTextClassifier',
+            state_training_jobs_mapping.algorithm_ids_to_job_ids)
+
+        # Check migration of an existing algorithm to a newer version.
+        mock_interaction_classifier_mapping = {
+            'TextInput': {
+                'algorithm_id': 'NewTextClassifier',
+                'algorithm_version': 2
+                },
+            }
+        with self.swap(
+            feconf,
+            'INTERACTION_CLASSIFIER_MAPPING',
+            mock_interaction_classifier_mapping):
+            classifier_services.migrate_state_training_jobs(
+                classifier_services.get_state_training_jobs_mapping(
+                    self.exp_id, 1, state_name))
+        next_job = classifier_services.fetch_next_job()
+        self.assertEqual(
+            mock_interaction_classifier_mapping[
+                'TextInput']['algorithm_version'],
+            next_job.algorithm_version
+        )
+
+    def test_reverted_exploration_maintains_classifier_model_mapping(self):
+        """Test if the classifier model mapping is maintained when an
+        exploration is reverted.
+        """
+
+        state_name = 'Home'
+        old_exploration = exp_fetchers.get_exploration_by_id(self.exp_id)
+        interaction_id = old_exploration.states[
+            state_name].interaction.id
+        algorithm_id = feconf.INTERACTION_CLASSIFIER_MAPPING[
+            interaction_id]['algorithm_id']
+        job = classifier_services.get_classifier_training_job(
+            self.exp_id, old_exploration.version, state_name,
+            algorithm_id)
+        change_list = [exp_domain.ExplorationChange({
+            'cmd': exp_domain.CMD_EDIT_EXPLORATION_PROPERTY,
+            'property_name': 'title',
+            'new_value': 'A new title'
+        })]
+        with self.swap(feconf, 'ENABLE_ML_CLASSIFIERS', True):
+            exp_services.update_exploration(
```

Comment:
I don't see any reverts going on here, so I don't get where the "exploration is reverted" part comes into play. I understand that you're creating a job for the reverted exploration but I'm not sure why you're calling that function -- it seems a bit out of context given the setup.

I suggest reading the code you're testing carefully again (as well as adjacent code) in order to understand what's going on, and then figure out appropriate tests for the behaviour that make sense when read through.

Reply:
Yes you are correct. I am not actually reverting the exploration. I am calling the `create_classifier_training_job_for_reverted_exploration` which would be called when an exploration is reverted and verifying the mappings using the job IDs like I said above.